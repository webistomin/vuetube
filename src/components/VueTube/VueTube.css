.vuetube {
  position: relative;
  cursor: pointer;
}

.vuetube__box {
  position: relative;
}

.vuetube__box::before {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1;
  box-sizing: content-box;
  width: 100%;
  height: 60px;
  padding-bottom: 50px;
  background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAADGCAYAAAAT+OqFAAAAdklEQVQoz42QQQ7AIAgEF/T/D+kbq/RWAlnQyyazA4aoAB4FsBSA/bFjuF1EOL7VbrIrBuusmrt4ZZORfb6ehbWdnRHEIiITaEUKa5EJqUakRSaEYBJSCY2dEstQY7AuxahwXFrvZmWl2rh4JZ07z9dLtesfNj5q0FU3A5ObbwAAAABJRU5ErkJggg==);
  background-repeat: repeat-x;
  background-position: top;
  transition: opacity 0.25s cubic-bezier(0, 0, 0.2, 1);
  content: '';
  pointer-events: none;
}

.vuetube__thumbnail {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.vuetube__image {
  width: inherit;
  height: inherit;
  object-fit: cover;
  object-position: center;
  vertical-align: top;
}

.vuetube__button {
  position: absolute;
  top: 50%;
  left: 50%;
  z-index: 2;
  margin: 0;
  padding: 0;
  overflow: visible;
  font-size: 100%;
  font-family: inherit;
  text-transform: none;
  background-color: transparent;
  border: none;
  transform: translate(-50%, -50%);
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  cursor: pointer;
  transition: opacity 0.25s cubic-bezier(0, 0, 0.2, 1), visibility 0.25s cubic-bezier(0, 0, 0.2, 1);
  -webkit-appearance: button;
}

.vuetube__button::-moz-focus-inner {
  padding: 0;
  border-style: none;
}

.vuetube__button:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.vuetube__icon {
  display: block;
  width: 68px;
  height: 48px;
}

.vuetube__icon-bg {
  transition: fill 0.1s cubic-bezier(0.4, 0, 1, 1), fill-opacity 0.1s cubic-bezier(0.4, 0, 1, 1);
  fill: #212121;
  fill-opacity: 0.8;
}

.vuetube:hover .vuetube__icon-bg {
  transition: fill 0.1s cubic-bezier(0, 0, 0.2, 1), fill-opacity 0.1s cubic-bezier(0, 0, 0.2, 1);
  fill: #f00;
  fill-opacity: 1;
}

.vuetube__icon-triangle {
  fill: #fff;
}

.vuetube__iframe {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border: 0;
}

.vuetube--played .vuetube__box::before {
  opacity: 0;
}

.vuetube--played .vuetube__button {
  visibility: hidden;
  opacity: 0;
}

.vuetube__button:focus:not(:focus-visible) {
  outline: none;
}
