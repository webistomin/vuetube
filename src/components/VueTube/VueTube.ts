import Vue, { PropType, VNode } from 'vue';
import { createWarmLink } from '@/utils/createWarmLink';
import { getStringifiedParams } from '@/utils/getStringifiedParams';
import { loadYoutubeAPI } from '@/utils/loadYoutubeAPI';
import { TImageLoading } from '@/types/image-loading';
import { TThumbnailResolution } from '@/types/thumbnail-resolution';

declare global {
  interface Window {
    YT: YT.Player;
    onYouTubeIframeAPIReady: () => void;
  }
}

export default /* #__PURE__ */ Vue.extend({
  name: 'VueTube',

  props: {
    /**
     * ID of YouTube video
     */
    videoId: {
      type: String as PropType<string>,
      required: true,
    },
    /**
     * Aspect ratio for iframe
     */
    ratio: {
      type: Number as PropType<number>,
      default: 16 / 9,
    },
    /**
     * Change video host to www.youtube.com
     * By default video loaded from https://www.youtube-nocookie.com
     */
    enableCookies: {
      type: Boolean as PropType<boolean>,
      default: false,
    },
    /**
     * Parameters that are available in the YouTube embedded player.
     * @link https://developers.google.com/youtube/player_parameters#Parameters
     */
    playerParameters: {
      type: Object as PropType<YT.PlayerVars>,
      default: () => ({ enablejsapi: 1 }),
    },
    /**
     * Disable warming up connections to origins that are in the critical path
     */
    disableWarming: {
      type: Boolean as PropType<boolean>,
      default: false,
    },
    /**
     * Disable webp thumbnail
     */
    disableWebp: {
      type: Boolean as PropType<boolean>,
      default: false,
    },
    /**
     * Alt attribute for image
     */
    imageAlt: {
      type: String as PropType<string>,
      default: '',
    },
    /**
     * Loading attribute for image
     */
    imageLoading: {
      type: String as PropType<TImageLoading>,
      default: 'lazy',
      validator: (value) => ['lazy', 'eager', 'auto'].indexOf(value) !== -1,
    },
    /**
     * Thumbnail from YouTube API
     * @link https://stackoverflow.com/questions/2068344/how-do-i-get-a-youtube-video-thumbnail-from-the-youtube-api
     */
    resolution: {
      type: String as PropType<TThumbnailResolution>,
      default: 'sddefault',
      validator: (value) => ['maxresdefault', 'sddefault', 'hqdefault'].indexOf(value) !== -1,
    },
    /**
     * Aria-label attribute for button
     */
    buttonLabel: {
      type: String as PropType<string>,
      default: 'Play video',
    },
    /**
     * Title attribute for iframe
     */
    iframeTitle: {
      type: String as PropType<string>,
      default: undefined,
    },
    /**
     * Allow attribute for iframe
     */
    iframeAllow: {
      type: String as PropType<string>,
      default: 'accelerometer;autoplay;encrypted-media;gyroscope;picture-in-picture',
    },
  },

  data(): {
    isConnectionWarmed: boolean;
    isPlayed: boolean;
    isLoaded: boolean;
    player: YT.Player | null
  } {
    return {
      /**
       * Is preconnect links already appended to the head
       */
      isConnectionWarmed: false,
      /**
       * Is video played
       */
      isPlayed: false,
      /**
       * Is video loaded
       */
      isLoaded: false,
      /**
       * Player instance
       */
      player: null,
    };
  },

  computed: {
    /**
     * Calculated video host
     */
    host(): string {
      const { enableCookies } = this;

      return enableCookies ? 'https://www.youtube.com' : 'https://www.youtube-nocookie.com';
    },
    /**
     * Calculate iframe url with params
     */
    iframeUrl(): string {
      const { playerParameters, host, videoId } = this;
      const DEFAULT_PARAMS: YT.PlayerVars = {
        autoplay: 1,
      };
      const concatParams = Object.assign(
        {},
        DEFAULT_PARAMS,
        playerParameters,
      );

      const serialisedParams = getStringifiedParams(concatParams);

      return `${host}/embed/${videoId}${serialisedParams}`;
    },
    /**
     * Wrapper component
     */
    boxComponent(): VNode {
      const {
        $createElement,
        ratio,
        isPlayed,
        iframeComponent,
        thumbnailComponent,
        playBtnComponent,
        warmConnections,
      } = this;

      const BOX_COMPONENT = $createElement(
        'div',
        {
          on: {
            '~pointerover': warmConnections,
          },
          class: [
            'vuetube',
            {
              'vuetube--played': isPlayed,
            },
          ],
          style: {
            '--vuetube-aspect-ratio': ratio,
          },
        },
        [isPlayed ? iframeComponent : thumbnailComponent, playBtnComponent],
      );

      return BOX_COMPONENT;
    },
    /**
     * Picture component
     */
    thumbnailComponent(): VNode | VNode[] {
      const {
        $scopedSlots,
        $createElement,
        resolution,
        videoId,
        disableWebp,
        imageAlt,
        imageLoading,
      } = this;

      const webp = `https://i.ytimg.com/vi_webp/${videoId}/${resolution}.webp`;
      const jpg = `https://i.ytimg.com/vi/${videoId}/${resolution}.jpg`;

      const THUMBNAIL_COMPONENT = $createElement(
        'picture',
        {
          class: [
            'vuetube__thumbnail',
          ],
        },
        [
          !disableWebp && $createElement(
            'source',
            {
              attrs: {
                srcset: webp,
                type: 'image/webp',
              },
            },
          ),
          $createElement(
            'source',
            {
              attrs: {
                srcset: jpg,
                type: 'image/jpeg',
              },
            },
          ),
          $createElement(
            'img',
            {
              class: [
                'vuetube__image',
              ],
              attrs: {
                src: jpg,
                alt: imageAlt,
                loading: imageLoading,
              },
            },
          ),
        ],
      );

      return $scopedSlots?.thumbnail?.({}) || THUMBNAIL_COMPONENT;
    },
    /**
     * Button component
     */
    playBtnComponent(): VNode {
      const {
        $createElement,
        buttonLabel,
        playVideo,
        playBtnIconComponent,
      } = this;

      const PLAY_BTN_COMPONENT = $createElement(
        'button',
        {
          class: [
            'vuetube__button',
          ],
          attrs: {
            type: 'button',
            'aria-label': buttonLabel,
          },
          on: {
            click: playVideo,
          },
        },
        [playBtnIconComponent],
      );

      return PLAY_BTN_COMPONENT;
    },
    /**
     * Icon component
     */
    playBtnIconComponent(): VNode | VNode[] {
      const { $createElement, $scopedSlots } = this;

      const PLAY_BTN_ICON_COMPONENT = $createElement(
        'svg',
        {
          attrs: {
            xmlns: 'http://www.w3.org/2000/svg',
            viewBox: '0 0 68 48',
            class: 'vuetube__icon',
            'aria-hidden': true,
            focusable: 'false',
          },
        },
        [
          $createElement(
            'path',
            {
              attrs: {
                class: 'vuetube__icon-bg',
                d: 'M66.52 7.74c-.78-2.93-2.49-5.41-5.42-6.19C55.79.13 34 0 34 0S12.21.13 6.9 1.55c-2.93.78-4.63 3.26-5.42 6.19C.06 13.05 0 24 0 24s.06 10.95 1.48 16.26c.78 2.93 2.49 5.41 5.42 6.19C12.21 47.87 34 48 34 48s21.79-.13 27.1-1.55c2.93-.78 4.64-3.26 5.42-6.19C67.94 34.95 68 24 68 24s-.06-10.95-1.48-16.26z',
              },
            },
          ),
          $createElement(
            'path',
            {
              attrs: {
                class: 'vuetube__icon-triangle',
                d: 'M45 24L27 14v20',
              },
            },
          ),
        ],
      );

      return $scopedSlots.icon?.({}) || PLAY_BTN_ICON_COMPONENT;
    },
    /**
     * Iframe component
     */
    iframeComponent(): VNode {
      const {
        $createElement,
        iframeTitle,
        onIframeLoad,
        iframeAllow,
        iframeUrl,
      } = this;

      const IFRAME_COMPONENT = $createElement(
        'iframe',
        {
          ref: 'iframe',
          class: [
            'vuetube__iframe',
          ],
          attrs: {
            src: iframeUrl,
            allow: iframeAllow,
            title: iframeTitle,
          },
          on: {
            load: onIframeLoad,
          },
        },
      );

      return IFRAME_COMPONENT;
    },
  },

  methods: {
    /**
     * Add preconnect links
     */
    warmConnections(): void {
      const { disableWarming, enableCookies, isConnectionWarmed } = this;

      if (disableWarming || isConnectionWarmed) {
        return;
      }

      if (enableCookies) {
        createWarmLink('preconnect', 'https://www.youtube.com');
        createWarmLink('preconnect', 'https://googleads.g.doubleclick.net');
      } else {
        createWarmLink('preconnect', 'https://www.youtube-nocookie.com');
      }

      createWarmLink('preconnect', 'https://fonts.gstatic.com', true);
      createWarmLink('preconnect', 'https://i.ytimg.com');
      createWarmLink('preconnect', 'https://google.com');
      createWarmLink('preconnect', 'https://yt3.ggpht.com');
      createWarmLink('preconnect', 'https://www.gstatic.com');

      this.isConnectionWarmed = true;
    },
    /**
     * Run video
     */
    playVideo(): void {
      this.isPlayed = true;
    },
    /**
     * Run after iframe has been loaded
     */
    onIframeLoad(): void {
      const {
        $refs,
        playerParameters,
        initAPI,
      } = this;

      this.isLoaded = true;

      const el = $refs.iframe as HTMLIFrameElement;
      el.focus();

      const SHOULD_LOAD_API = playerParameters.enablejsapi;

      if (SHOULD_LOAD_API) {
        initAPI();
      }
    },

    initAPI() {
      const {
        $refs,
        videoId,
        playerParameters,
      } = this;

      const el = $refs.iframe as HTMLIFrameElement;

      window.onYouTubeIframeAPIReady = () => {
        const player = new window.YT.Player(el, {
          videoId,
          playerVars: playerParameters,
          events: {
            onReady: ($event) => this.$emit('player:ready', $event),
            onStateChange: ($event) => this.$emit('player:statechange', $event),
            onPlaybackQualityChange: ($event) => this.$emit('player:playbackqualitychange', $event),
            onPlaybackRateChange: ($event) => this.$emit('player:playbackratechange', $event),
          },
        });

        this.player = player;
      };

      loadYoutubeAPI();
    },
  },
  /**
   * Render component
   */
  render(): VNode {
    return this.boxComponent;
  },
});
