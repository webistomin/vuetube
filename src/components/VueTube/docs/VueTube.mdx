import { Meta, Story, Preview } from "@storybook/addon-docs/blocks"
import VueTube from "../VueTube"

<Meta title='Components/VueTube' component={ VueTube } />

# VueTube

## Preview

<Preview className="custom-preview">
  <Story id='components-vuetube--default' />
</Preview>

## Example

### Props

<table>
  <tr>
    <td>
      Prop
    </td>
    <td>
      Type
    </td>
    <td>
      Description
    </td>
    <td>
      Default value
    </td>
  </tr>
  <tr>
    <td>
      <code>videoId</code>
    </td>
    <td>
      <code>string</code>
    </td>
    <td>
      The ID of YouTube video (required)
    </td>
    <td>
      -
    </td>
  </tr>
  <tr>
    <td>
      <code>isPlaylist</code>
    </td>
    <td>
      <code>boolean</code>
    </td>
    <td>
      Should embed a playlist of several videos
    </td>
    <td>
      <code>false</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>aspectRatio</code>
    </td>
    <td>
      <code>number</code>
    </td>
    <td>
      The aspect ratio for iframe
    </td>
    <td>
      <code>16 / 9</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>enableCookies</code>
    </td>
    <td>
      <code>boolean</code>
    </td>
    <td>
      Change video host to <code>www.youtube.com</code>. By default, video loaded from <code>https://www.youtube-nocookie.com</code>.
    </td>
    <td>
      <code>false</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>playerVars</code>
    </td>
    <td>
      <code>object</code>
    </td>
    <td>
      <a href="https://developers.google.com/youtube/player_parameters#Parameters">Parameters</a> that are available in the YouTube embedded player.
    </td>
    <td>
      <code>{}</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>disableWarming</code>
    </td>
    <td>
      <code>boolean</code>
    </td>
    <td>
      Disable warming up connections to origins that are in the critical path on component hover.
    </td>
    <td>
      <code>false</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>disableWebp</code>
    </td>
    <td>
      <code>boolean</code>
    </td>
    <td>
      Disable webp thumbnail.
    </td>
    <td>
      <code>false</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>imageAlt</code>
    </td>
    <td>
      <code>string</code>
    </td>
    <td>
      Alt attribute for image
    </td>
    <td>
      <code>''</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>imageLoading</code>
    </td>
    <td>
      <code>string</code>
    </td>
    <td>
      <a href="https://caniuse.com/loading-lazy-attr">Loading attribute</a> for image
    </td>
    <td>
      <code>'lazy'</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>resolution</code>
    </td>
    <td>
      <code>string</code>
    </td>
    <td>
      <a href="https://stackoverflow.com/a/18400445/13374604">Thumbnail resolution</a> from YouTube API.
    </td>
    <td>
      <code>'sddefault'</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>buttonLabel</code>
    </td>
    <td>
      <code>string</code>
    </td>
    <td>
      Aria-label attribute for button
    </td>
    <td>
      <code>'Play video'</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>iframeTitle</code>
    </td>
    <td>
      <code>string</code>
    </td>
    <td>
      Title attribute for iframe
    </td>
    <td>
      <code>''</code>
    </td>
  </tr>
  <tr>
    <td>
      <code>iframeAllow</code>
    </td>
    <td>
      <code>string</code>
    </td>
    <td>
      Allow attribute for iframe
    </td>
    <td>
      <code>'accelerometer;autoplay;encrypted-media;gyroscope;picture-in-picture'</code>
    </td>
  </tr>
</table>

### Events

<table>
  <tr>
    <td>Event name</td>
    <td>Description</td>
  </tr>
  <tr>
    <td><code>player:play</code></td>
    <td>The user clicked on the play button</td>
  </tr>
  <tr>
    <td><code>player:load</code></td>
    <td>Iframe has been loaded</td>
  </tr>
  <tr>
    <td><code>player:ready</code></td>
    <td>This event fires whenever a player has finished loading and is ready to begin receiving API calls. <br/><br/> <strong>Make sure you pass <code>enablejsapi: 1</code> to the <code>playerVars</code> props object</strong> <br/><br/> This event is similar to <code>onReady</code> event from <a href="https://developers.google.com/youtube/iframe_api_reference#Events">Youtube API documentation</a>.</td>
  </tr>
  <tr>
    <td><code>player:statechange</code></td>
    <td>This event fires whenever the player's state changes. <br/><br/> <strong>Make sure you pass <code>enablejsapi: 1</code> to the <code>playerVars</code> props object</strong> <br/><br/> This event is similar to <code>onStateChange</code> event from <a href="https://developers.google.com/youtube/iframe_api_reference#Events">Youtube API documentation</a>.</td>
  </tr>
  <tr>
    <td><code>player:playbackqualitychange</code></td>
    <td>This event fires whenever the video playback quality changes. <br/><br/> <strong>Make sure you pass <code>enablejsapi: 1</code> to the <code>playerVars</code> props object</strong> <br/><br/> This event is similar to <code>onPlaybackQualityChange</code> event from <a href="https://developers.google.com/youtube/iframe_api_reference#Events">Youtube API documentation</a>.</td>
  </tr>
  <tr>
    <td><code>player:playbackratechange</code></td>
    <td>This event fires whenever the video playback rate changes.  <br/><br/> <strong>Make sure you pass <code>enablejsapi: 1</code> to the <code>playerVars</code> props object</strong> <br/><br/> This event is similar to <code>onPlaybackRateChange</code> event from <a href="https://developers.google.com/youtube/iframe_api_reference#Events">Youtube API documentation</a>.</td>
  </tr>
  <tr>
    <td><code>player:error</code></td>
    <td>This event fires if an error occurs in the player. <br/><br/> <strong>Make sure you pass <code>enablejsapi: 1</code> to the <code>playerVars</code> props object</strong> <br/><br/> This event is similar to <code>onError</code> event from <a href="https://developers.google.com/youtube/iframe_api_reference#Events">Youtube API documentation</a>.</td>
  </tr>
  <tr>
    <td><code>player:apichange</code></td>
    <td>This event is fired to indicate that the player has loaded (or unloaded) a module with exposed API methods. <br/><br/> <strong>Make sure you pass <code>enablejsapi: 1</code> to the <code>playerVars</code> props object</strong> <br/><br/> This event is similar to <code>onApiChange</code> event from <a href="https://developers.google.com/youtube/iframe_api_reference#Events">Youtube API documentation</a>.</td>
  </tr>
</table>

## FAQ üóØÔ∏è

### How to get access to player API?

You need to pass <code>enablejsapi: 1</code> to the <code>playerVars</code> props object. Then subscribe on <code>player:ready</code> event.

```vue
  <template>
    <vue-tube @player:ready="onPlayerReady"></vue-tube>
  </template>

  <script>
    export default {
      data() {
        return {
          playerInstance: null,
        }
      },

      methods: {
        onPlayerReady(event) {
          this.playerInstance = event.target
        }
      }
    }
  </script>
```

Now you have a player instance in the `playerInstance`. You can do whatever you want with video.

`this.playerInstance.playVideo()`, `this.playerInstance.pauseVideo()` etc. All methods can be found [here](https://developers.google.com/youtube/iframe_api_reference#Playback_controls).

### How to use my play button?

You can pass your button through the `icon` slot.

```vue
  <template>
    <vue-tube>
      <template #icon>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="48"
          height="48"
          viewBox="0 0 24 24"
          aria-hidden="true"
          focusable="false"
        >
          <path
            fill="#fff"
            d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12zm-3 17v-10l9 5.146-9 4.854z"
          />
        </svg>
      </template>
    </vue-tube>
  </template>
```

### How to use my thumbnail?

You can pass your thumbnail through the `thumbnail` slot.

```vue
  <template>
    <vue-tube>
      <template #thumbnail>
        <img src="#" alt="">
      </template>
    </vue-tube>
  </template>
```

