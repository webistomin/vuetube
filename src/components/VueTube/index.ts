import VueTube from './VueTube';

export default VueTube;
