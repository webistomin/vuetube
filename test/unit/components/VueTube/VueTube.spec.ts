import { mount } from '@vue/test-utils';
import VueTube from '@/components/VueTube';

describe('VueTube.vue', () => {
  const factory = (propsData = {}, options = {}) => mount(VueTube, {
    propsData: {
      videoId: 'dQw4w9WgXcQ',
      isPlaylist: false,
      aspectRatio: 16 / 9,
      enableCookies: false,
      playerVars: {},
      disableWarming: false,
      disableWebp: false,
      imageAlt: 'Rick Astley',
      imageLoading: 'lazy',
      resolution: 'sddefault',
      buttonLabel: 'Play video',
      iframeTitle: 'Rick Astley - Never Gonna Give You Up (Video)',
      iframeAllow: 'accelerometer;autoplay;encrypted-media;gyroscope;picture-in-picture',
      ...propsData,
    },
    ...options,
  });

  it('set correct image urls and types, when using webp', () => {
    const wrapper = factory();
    const sources = wrapper.findAll('source');
    const image = wrapper.find('img');
    const sourceWebp = sources.at(0).element as HTMLSourceElement;
    const sourceJpeg = sources.at(1).element as HTMLSourceElement;
    const sourceImg = image.element as HTMLImageElement;

    const WEBP_SRCSET = 'https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp';
    const JPEG_SRCSET = 'https://i.ytimg.com/vi/dQw4w9WgXcQ/sddefault.jpg';
    const WEBP_TYPE = 'image/webp';
    const JPEG_TYPE = 'image/jpeg';

    expect(sources.length).toEqual(2);
    expect(sourceWebp.srcset).toEqual(WEBP_SRCSET);
    expect(sourceWebp.type).toEqual(WEBP_TYPE);
    expect(sourceJpeg.srcset).toEqual(JPEG_SRCSET);
    expect(sourceJpeg.type).toEqual(JPEG_TYPE);
    expect(sourceImg.src).toEqual(JPEG_SRCSET);
  });

  it('set correct image urls and types, when webp is disabled', () => {
    const wrapper = factory(
      {
        disableWebp: true,
      },
    );
    const sources = wrapper.findAll('source');
    const image = wrapper.find('img');
    const sourceJpeg = sources.at(0).element as HTMLSourceElement;
    const sourceImg = image.element as HTMLImageElement;

    const JPEG_SRCSET = 'https://i.ytimg.com/vi/dQw4w9WgXcQ/sddefault.jpg';
    const JPEG_TYPE = 'image/jpeg';

    expect(sources.length).toEqual(1);
    expect(sourceJpeg.srcset).toEqual(JPEG_SRCSET);
    expect(sourceJpeg.type).toEqual(JPEG_TYPE);
    expect(sourceImg.src).toEqual(JPEG_SRCSET);
  });

  it('load iframe on component click', async () => {
    const wrapper = factory();
    await wrapper.trigger('click');

    const iframe = wrapper.find('iframe');

    expect(iframe.exists()).toBe(true);
  });

  it('load iframe on button click', async () => {
    const wrapper = factory();
    const button = wrapper.find('button');
    await button.trigger('click');

    const iframe = wrapper.find('iframe');

    expect(iframe.exists()).toBe(true);
  });

  it('match snapshot when iframe is loaded', async () => {
    const wrapper = factory();
    await wrapper.trigger('click');

    const html = wrapper.html();

    expect(html).toMatchSnapshot();
  });

  it('match snapshot when using custom button icon', () => {
    const wrapper = factory({}, {
      scopedSlots: {
        icon: '<svg class="my-custom-icon"></svg>',
      },
    });
    const html = wrapper.html();

    expect(html).toMatchSnapshot();
  });

  it('match snapshot when using custom thumbnail', () => {
    const wrapper = factory({}, {
      scopedSlots: {
        thumbnail: '<img src="#" alt="" class="my-custom-thumbnail" />',
      },
    });
    const html = wrapper.html();

    expect(html).toMatchSnapshot();
  });
});
