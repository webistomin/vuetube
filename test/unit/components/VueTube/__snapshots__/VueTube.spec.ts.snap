// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueTube.vue match snapshot when iframe is loaded 1`] = `
<div class="vuetube vuetube--played">
  <div class="vuetube__box" style="padding-bottom: 56.25%;"><iframe src="https://www.youtube-nocookie.com/embed/dQw4w9WgXcQ?autoplay=1" allow="accelerometer;autoplay;encrypted-media;gyroscope;picture-in-picture" title="Rick Astley - Never Gonna Give You Up (Video)" allowfullscreen="allowfullscreen" class="vuetube__iframe"></iframe><button type="button" aria-label="Play video" class="vuetube__button"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 68 48" class="vuetube__icon" aria-hidden="true" focusable="false">
        <path class="vuetube__icon-bg" d="M66.52 7.74c-.78-2.93-2.49-5.41-5.42-6.19C55.79.13 34 0 34 0S12.21.13 6.9 1.55c-2.93.78-4.63 3.26-5.42 6.19C.06 13.05 0 24 0 24s.06 10.95 1.48 16.26c.78 2.93 2.49 5.41 5.42 6.19C12.21 47.87 34 48 34 48s21.79-.13 27.1-1.55c2.93-.78 4.64-3.26 5.42-6.19C67.94 34.95 68 24 68 24s-.06-10.95-1.48-16.26z"></path>
        <path class="vuetube__icon-triangle" d="M45 24L27 14v20"></path>
      </svg></button></div>
</div>
`;

exports[`VueTube.vue match snapshot when using custom button icon 1`] = `
<div class="vuetube">
  <div class="vuetube__box" style="padding-bottom: 56.25%;">
    <picture class="vuetube__thumbnail">
      <source srcset="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp" type="image/webp">
      <source srcset="https://i.ytimg.com/vi/dQw4w9WgXcQ/sddefault.jpg" type="image/jpeg"><img src="https://i.ytimg.com/vi/dQw4w9WgXcQ/sddefault.jpg" alt="Rick Astley" loading="lazy" class="vuetube__image">
    </picture><button type="button" aria-label="Play video" class="vuetube__button"><svg class="my-custom-icon"></svg></button>
  </div>
</div>
`;

exports[`VueTube.vue match snapshot when using custom thumbnail 1`] = `
<div class="vuetube">
  <div class="vuetube__box" style="padding-bottom: 56.25%;"><img src="#" alt="" class="my-custom-thumbnail"><button type="button" aria-label="Play video" class="vuetube__button"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 68 48" class="vuetube__icon" aria-hidden="true" focusable="false">
        <path class="vuetube__icon-bg" d="M66.52 7.74c-.78-2.93-2.49-5.41-5.42-6.19C55.79.13 34 0 34 0S12.21.13 6.9 1.55c-2.93.78-4.63 3.26-5.42 6.19C.06 13.05 0 24 0 24s.06 10.95 1.48 16.26c.78 2.93 2.49 5.41 5.42 6.19C12.21 47.87 34 48 34 48s21.79-.13 27.1-1.55c2.93-.78 4.64-3.26 5.42-6.19C67.94 34.95 68 24 68 24s-.06-10.95-1.48-16.26z"></path>
        <path class="vuetube__icon-triangle" d="M45 24L27 14v20"></path>
      </svg></button></div>
</div>
`;
